lotus-soup/rfwp/diffs.go

```diff
@@ -0,0 +1,295 @@
+package rfwp
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"sort"
+	"sync"
+
+	"github.com/filecoin-project/oni/lotus-soup/testkit"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/abi/big"
+)
+
+type ChainState struct {
+	sync.Mutex
+
+	PrevHeight abi.ChainEpoch
+	DiffHeight map[string]map[string]map[abi.ChainEpoch]big.Int  // height -> value
+	DiffValue  map[string]map[string]map[string][]abi.ChainEpoch // value -> []height
+	DiffCmp    map[string]map[string]map[string][]abi.ChainEpoch // difference (height, height-1) -> []height
+	valueTypes []string
+}
+
+func NewChainState() *ChainState {
+	cs := &ChainState{}
+	cs.PrevHeight = abi.ChainEpoch(-1)
+	cs.DiffHeight = make(map[string]map[string]map[abi.ChainEpoch]big.Int) // height -> value
+	cs.DiffValue = make(map[string]map[string]map[string][]abi.ChainEpoch) // value -> []height
+	cs.DiffCmp = make(map[string]map[string]map[string][]abi.ChainEpoch)   // difference (height, height-1) -> []height
+	cs.valueTypes = []string{"MinerPower", "CommittedBytes", "ProvingBytes", "Balance", "PreCommitDeposits", "LockedFunds", "AvailableFunds", "WorkerBalance", "MarketEscrow", "MarketLocked", "Faults", "ProvenSectors", "Recoveries"}
+	return cs
+}
+
+var (
+	cs *ChainState
+)
+
+func init() {
+	cs = NewChainState()
+}
+
+func printDiff(t *testkit.TestEnvironment, mi *MinerInfo, height abi.ChainEpoch) {
+	maddr := mi.MinerAddr.String()
+	filename := fmt.Sprintf("%s%cdiff-%s-%d", t.TestOutputsPath, os.PathSeparator, maddr, height)
+
+	f, err := os.Create(filename)
+	if err != nil {
+		panic(err)
+	}
+	defer f.Close()
+
+	w := bufio.NewWriter(f)
+	defer w.Flush()
+
+	keys := make([]string, 0, len(cs.DiffCmp[maddr]))
+	for k := range cs.DiffCmp[maddr] {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	fmt.Fprintln(w, "=====", maddr, "=====")
+	for i, valueName := range keys {
+		fmt.Fprintln(w, toCharStr(i), "=====", valueName, "=====")
+		if len(cs.DiffCmp[maddr][valueName]) > 0 {
+			fmt.Fprintf(w, "%s diff of             |\n", toCharStr(i))
+		}
+
+		for difference, heights := range cs.DiffCmp[maddr][valueName] {
+			fmt.Fprintf(w, "%s diff of %30v at heights %v\n", toCharStr(i), difference, heights)
+		}
+	}
+}
+
+func recordDiff(mi *MinerInfo, ps *ProvingInfoState, height abi.ChainEpoch) {
```

Comment:
I think we might end up reusing some of this diff logic for our test-vector-driven testing.